// @opuscapita/react-datetime/lib/es/year-month-picker/year-month-picker.component.js
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import PropTypes from 'prop-types';
import LocaleUtils from 'react-day-picker/moment';
import { FormControl } from 'react-bootstrap'; // App imports

import './year-month-picker.scss';

var YearMonthPicker =
/*#__PURE__*/
function (_React$Component) {
  _inheritsLoose(YearMonthPicker, _React$Component);

  function YearMonthPicker(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;

    _defineProperty(_assertThisInitialized(_this), "onChange", function (event) {
      var onChange = _this.props.onChange;
      var _event$target$form = event.target.form,
          year = _event$target$form.year,
          month = _event$target$form.month;
      onChange(new Date(year.value, month.value));
    });

    _defineProperty(_assertThisInitialized(_this), "getYears", function (date) {
      if (!date) return 0;
      return date.getYear();
    });

    _defineProperty(_assertThisInitialized(_this), "getMonths", function (date) {
      if (!date) return 0;
      return date.getMonth();
    });

    _this.months = [];
    _this.years = [];
    var currentYear = new Date().getFullYear();
    var fromMonth = new Date(currentYear - 10, 0);
    var toMonth = new Date(currentYear + 10, 11);
    _this.months = LocaleUtils.getMonths(props.locale);

    for (var i = fromMonth.getFullYear(); i <= toMonth.getFullYear(); i += 1) {
      _this.years.push(i);
    }

    return _this;
  }
  /**
   * On select box change
   * @param event
   */


  var _proto = YearMonthPicker.prototype;

  _proto.render = function render() {
    var date = this.props.date;
    var year = date && date.getFullYear ? date.getFullYear() : null;
    var month = date && date.getMonth ? date.getMonth() : null;
    return React.createElement("form", {
      className: "DayPicker-Caption oc-year-month-picker-container"
    }, React.createElement(FormControl, {
      name: "year",
      componentClass: "select",
      value: year,
      onChange: this.onChange,
      className: "year"
    }, this.years.map(function (y) {
      return React.createElement("option", {
        key: "year-" + y,
        value: y
      }, y);
    })), React.createElement(FormControl, {
      name: "month",
      componentClass: "select",
      value: month,
      onChange: this.onChange,
      className: "month"
    }, this.months.map(function (m, index) {
      return React.createElement("option", {
        key: "month-" + m,
        value: index
      }, m);
    })));
  };

  return YearMonthPicker;
}(React.Component);

_defineProperty(YearMonthPicker, "defaultProps", {
  date: null,
  locale: 'en-GB'
});

export { YearMonthPicker as default };
