// @opuscapita/react-datetime/lib/cjs/navbar/navbar.component.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _keys = require("../keys");

require("./navbar.component.scss");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Navbar =
/*#__PURE__*/
function (_Component) {
  _inheritsLoose(Navbar, _Component);

  function Navbar() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;

    _defineProperty(_assertThisInitialized(_this), "handleNextClick", function () {
      var onNextClick = _this.props.onNextClick;
      if (onNextClick) onNextClick();
    });

    _defineProperty(_assertThisInitialized(_this), "handlePreviousClick", function () {
      var onPreviousClick = _this.props.onPreviousClick;
      if (onPreviousClick) onPreviousClick();
    });

    _defineProperty(_assertThisInitialized(_this), "handleNextKeyDown", function (e) {
      if (e.keyCode !== _keys.ENTER && e.keyCode !== _keys.SPACE) return;
      e.preventDefault();

      _this.handleNextClick();
    });

    _defineProperty(_assertThisInitialized(_this), "handlePreviousKeyDown", function (e) {
      if (e.keyCode !== _keys.ENTER && e.keyCode !== _keys.SPACE) return;
      e.preventDefault();

      _this.handlePreviousClick();
    });

    return _this;
  }

  var _proto = Navbar.prototype;

  _proto.shouldComponentUpdate = function shouldComponentUpdate(nextProps) {
    var _this$props = this.props,
        labels = _this$props.labels,
        dir = _this$props.dir,
        showPreviousButton = _this$props.showPreviousButton,
        showNextButton = _this$props.showNextButton;
    return nextProps.labels !== labels || nextProps.dir !== dir || showPreviousButton !== nextProps.showPreviousButton || showNextButton !== nextProps.showNextButton;
  };

  _proto.render = function render() {
    var _this$props2 = this.props,
        showPreviousButton = _this$props2.showPreviousButton,
        showNextButton = _this$props2.showNextButton,
        labels = _this$props2.labels,
        dir = _this$props2.dir;
    var previousClickHandler;
    var nextClickHandler;
    var previousKeyDownHandler;
    var nextKeyDownHandler;
    var shouldShowPrevious;
    var shouldShowNext;

    if (dir === 'rtl') {
      previousClickHandler = this.handleNextClick;
      nextClickHandler = this.handlePreviousClick;
      previousKeyDownHandler = this.handleNextKeyDown;
      nextKeyDownHandler = this.handlePreviousKeyDown;
      shouldShowNext = showPreviousButton;
      shouldShowPrevious = showNextButton;
    } else {
      previousClickHandler = this.handlePreviousClick;
      nextClickHandler = this.handleNextClick;
      previousKeyDownHandler = this.handlePreviousKeyDown;
      nextKeyDownHandler = this.handleNextKeyDown;
      shouldShowNext = showNextButton;
      shouldShowPrevious = showPreviousButton;
    }

    var previousClassName = shouldShowPrevious ? 'DayPicker-NavButton prev' : 'DayPicker-NavButton--interactionDisabled';
    var nextClassName = shouldShowNext ? 'DayPicker-NavButton next' : 'DayPicker-NavButton--interactionDisabled';

    var previousButton = _react["default"].createElement("span", {
      tabIndex: "0",
      role: "button",
      "aria-label": labels.previousMonth,
      key: "previous",
      className: previousClassName,
      onKeyDown: shouldShowPrevious ? previousKeyDownHandler : undefined,
      onClick: shouldShowPrevious ? previousClickHandler : undefined
    }, "\u25C0");

    var nextButton = _react["default"].createElement("span", {
      tabIndex: "0",
      role: "button",
      "aria-label": labels.nextMonth,
      key: "right",
      className: nextClassName,
      onKeyDown: shouldShowNext ? nextKeyDownHandler : undefined,
      onClick: shouldShowNext ? nextClickHandler : undefined
    }, "\u25B6");

    return _react["default"].createElement("div", {
      className: "DayPicker-NavBar"
    }, dir === 'rtl' ? [nextButton, previousButton] : [previousButton, nextButton]);
  };

  return Navbar;
}(_react.Component);

exports["default"] = Navbar;

_defineProperty(Navbar, "defaultProps", {
  dir: 'ltr',
  labels: {
    previousMonth: 'Previous Month',
    nextMonth: 'Next Month'
  },
  showPreviousButton: true,
  showNextButton: true,
  onPreviousClick: null,
  onNextClick: null
});
