// @opuscapita/react-datetime/lib/cjs/date-input.component.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactBootstrap = require("react-bootstrap");

var _moment = _interopRequireDefault(require("moment"));

var _reactDayPicker = _interopRequireWildcard(require("react-day-picker"));

var _moment2 = _interopRequireDefault(require("react-day-picker/moment"));

var _reactTether = _interopRequireDefault(require("react-tether"));

require("react-day-picker/lib/style.css");

var _timePicker = _interopRequireDefault(require("./time-picker/time-picker.component"));

var _yearMonthPicker = _interopRequireDefault(require("./year-month-picker/year-month-picker.component"));

var _navbar = _interopRequireDefault(require("./navbar/navbar.component"));

require("./date-input.scss");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// Date formats used by the component (mainly by the getDate method)
var FORMATS = {
  UTC: 'UTC',
  PRETTY_DATE: 'PRETTY_DATE',
  DATE_OBJECT: 'DATE_OBJECT'
}; // Used in getTetherComponentAttachmentLocation fn

var DATETIME_POPUP_HEIGHT = 200;
var classPrefix = 'oc-datetime';

var DateInput =
/*#__PURE__*/
function (_React$Component) {
  _inheritsLoose(DateInput, _React$Component);

  DateInput.getDerivedStateFromProps = function getDerivedStateFromProps(props, state) {
    var formatDate = props.formatDate,
        value = props.value;

    if (!state.showOverlay && value !== state.lastValue) {
      var momentDate = _moment["default"].utc(value, _moment["default"].ISO_8601);

      var inputDate = formatDate ? formatDate(value) : DateInput.getDate(momentDate, FORMATS.PRETTY_DATE, props.dateFormat);
      return {
        lastValue: value,
        selectedDay: DateInput.getDate(momentDate, FORMATS.DATE_OBJECT),
        showOverlay: props.showOverlay || state.showOverlay,
        inputDate: inputDate
      };
    }

    return null;
  }
  /**
   * Converts given date into wanted type (string/date object)
   * @param date - {string, moment object}
   * @param type - {string, date object} type of the return value
   * @param dateFormat {string} date format, defaults to 'M/D/YYYY'
   * ('M/D/YYYY' h:mm when using DateTime)
   * * @returns {string, date}
   */
  ;

  DateInput.getDate = function getDate(date, type, dateFormat) {
    var momentDate = typeof date === 'string' ? _moment["default"].utc(date, dateFormat) : date;
    if (!momentDate.isValid() || !date) return '';

    switch (type) {
      case FORMATS.PRETTY_DATE:
        return DateInput.removeInvisibleChars(momentDate.format(dateFormat));

      case FORMATS.UTC:
        return DateInput.removeInvisibleChars(momentDate.toISOString());

      case FORMATS.DATE_OBJECT:
      default:
        // UTC day might differ from local day, therefore UTC offset
        // must be discounted.
        return new Date((0, _moment["default"])(momentDate.format('L'), 'L'));
    }
  };

  function DateInput(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;

    _defineProperty(_assertThisInitialized(_this), "onDocumentClick", function (e) {
      if (!_this.calendarContainer) return;
      var showOverlay = _this.state.showOverlay; // Closes overlay if user clicks outside the calendar (and input field)

      if (!_this.calendarContainer.contains(e.target) && showOverlay && e.target !== _this.input) {
        _this.closeOverlay();

        document.removeEventListener('click', _this.onDocumentClick);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "getFirstDayOfWeek", function () {
      var locale = _this.props.locale;
      return _moment["default"].localeData(locale).firstDayOfWeek();
    });

    _defineProperty(_assertThisInitialized(_this), "getTetherComponentAttachmentLocation", function () {
      var time = _this.props.time;

      var inputDimensions = _this.input && _this.input.getBoundingClientRect(); // Popup will open below the input by default


      var attachment = 'top center';

      if (inputDimensions) {
        /* If there's time inputs present, the popup will be slightly taller. Height has to be
        hard coded, because we cannot determine the height of the popup before we have opened it */
        var popupHeight = time ? DATETIME_POPUP_HEIGHT + 50 : DATETIME_POPUP_HEIGHT;
        var popupBottomY = popupHeight + inputDimensions.height + inputDimensions.y;
        var windowHeight = window.innerHeight; // Popup has no space to open below the input, so..

        if (windowHeight < popupBottomY) attachment = 'bottom center';
      }

      return attachment;
    });

    _defineProperty(_assertThisInitialized(_this), "handleInputFocus", function (e) {
      var onFocus = _this.props.inputProps.onFocus;

      _this.setState({
        showOverlay: true
      }, function () {
        // Delays the execution so that the dayPicker opens before selecting a day
        setTimeout(function () {
          var _this$state = _this.state,
              showOverlay = _this$state.showOverlay,
              selectedDay = _this$state.selectedDay;
          if (!showOverlay && _this.dayPicker && selectedDay) _this.dayPicker.showMonth(selectedDay);
        });
      });

      document.addEventListener('click', _this.onDocumentClick);
      if (onFocus) onFocus(e);
    });

    _defineProperty(_assertThisInitialized(_this), "closeOverlay", function (e) {
      _this.setState({
        showOverlay: false
      }, function () {
        var showOverlay = _this.state.showOverlay;
        var onBlur = _this.props.inputProps.onBlur;
        if (showOverlay) _this.input.focus();
        if (onBlur) onBlur(e);
      });
    });

    _defineProperty(_assertThisInitialized(_this), "handleInputChange", function (e) {
      var inputDate = e.target.value;
      var _this$props = _this.props,
          dateFormat = _this$props.dateFormat,
          inputProps = _this$props.inputProps,
          onChange = _this$props.onChange;

      _this.setState({
        inputDate: inputDate
      }); // This fires only if the new date is valid in given format


      if (_moment["default"].utc(inputDate, dateFormat).isValid() && _this.isValidFormat(inputDate)) {
        _this.setState({
          selectedDay: DateInput.getDate(inputDate, FORMATS.DATE_OBJECT, dateFormat)
        }, function () {
          var selectedDay = _this.state.selectedDay; // If dayPicker is open, we will show the correct month

          if (_this.dayPicker) _this.dayPicker.showMonth(selectedDay);
        });

        if (inputProps.onChange) {
          inputProps.onChange(DateInput.removeInvisibleChars(inputDate));
        } else {
          onChange(DateInput.getDate(inputDate, FORMATS.UTC, dateFormat));
        }
      } else {
        // If the value is invalid we reset the model value
        onChange(null);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleInputBlur", function (e) {
      var onBlur = _this.props.inputProps.onBlur;

      _this.prettifyInputDate(); // We want to close the overlay on blur, unless it was caused by a click on the calendar
      // overlay


      if (!_this.mouseClickedOnContainer) {
        _this.setState({
          showOverlay: false
        });
      }

      _this.mouseClickedOnContainer = false;
      if (onBlur) onBlur(e);
    });

    _defineProperty(_assertThisInitialized(_this), "handleDayClick", function (day, modifiers) {
      if (modifiers === void 0) {
        modifiers = {};
      }

      if (modifiers.disabled) return;
      var _this$props2 = _this.props,
          dateFormat = _this$props2.dateFormat,
          formatDate = _this$props2.formatDate,
          value = _this$props2.value,
          time = _this$props2.time,
          onChange = _this$props2.onChange,
          onDayClick = _this$props2.onDayClick; // UTC day might differ from local date therefore UTC offset must be discounted.

      var momentDate = _moment["default"].utc((0, _moment["default"])(day).format('L'), 'L');

      var timeAdjustedDate = null;
      var currentMomentDate = (0, _moment["default"])(value, _moment["default"].ISO_8601).utc();
      var currentHours = currentMomentDate.get('hour');
      var currentMinutes = currentMomentDate.get('minute');

      if (time) {
        // Set current (previously selected) time to newly picked date
        timeAdjustedDate = momentDate.set('hour', currentHours).set('minute', currentMinutes);
      } else {
        // If we don't need to bother ourselves with an exact time,
        // we can set time to T00:00:00.000Z
        timeAdjustedDate = momentDate.startOf('day');
      }

      var inputDate = formatDate ? formatDate(timeAdjustedDate) : DateInput.getDate(timeAdjustedDate, FORMATS.PRETTY_DATE, dateFormat);

      _this.setState({
        selectedDay: day,
        showOverlay: false,
        inputDate: inputDate
      }, function () {
        onChange(DateInput.getDate(timeAdjustedDate, FORMATS.UTC, dateFormat));

        _this.input.blur();
      });

      onDayClick(day, modifiers);
    });

    _defineProperty(_assertThisInitialized(_this), "handleTimePickerChange", function (newTime) {
      var _this$props3 = _this.props,
          dateFormat = _this$props3.dateFormat,
          formatDate = _this$props3.formatDate,
          value = _this$props3.value,
          onChange = _this$props3.onChange;

      var momentDate = _moment["default"].utc(value);

      momentDate = momentDate.hour(newTime.hour);
      momentDate = momentDate.minutes(newTime.minute);
      var inputDate = formatDate ? formatDate(value) : DateInput.getDate(momentDate, FORMATS.PRETTY_DATE, dateFormat);

      _this.setState({
        inputDate: inputDate
      }, function () {
        onChange(DateInput.getDate(momentDate, FORMATS.UTC, dateFormat));
      });
    });

    _defineProperty(_assertThisInitialized(_this), "handleYearMonthChange", function (val) {
      var _this$props4 = _this.props,
          value = _this$props4.value,
          dateFormat = _this$props4.dateFormat,
          formatDate = _this$props4.formatDate,
          onChange = _this$props4.onChange;
      var momentDate = value ? _moment["default"].utc(value, _moment["default"].ISO_8601) : _moment["default"].utc();
      momentDate.year(val.getFullYear()).month(val.getMonth());
      var inputDate = formatDate ? formatDate(value) : DateInput.getDate(momentDate, FORMATS.PRETTY_DATE, dateFormat);

      _this.setState({
        inputDate: inputDate,
        selectedDay: DateInput.getDate(momentDate, FORMATS.DATE_OBJECT, dateFormat),
        dayPickerVisibleMonth: val
      }, function () {
        onChange(DateInput.getDate(momentDate, FORMATS.UTC, dateFormat));
      });
    });

    _defineProperty(_assertThisInitialized(_this), "handleOnOverlayMouseDown", function (e) {
      if (_this.calendarContainer.contains(e.target)) {
        _this.mouseClickedOnContainer = true;
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleClearClick", function () {
      var onChange = _this.props.onChange;
      if (!onChange) throw new TypeError('react-datetime: onChange callback is not set');
      onChange('');
    });

    _defineProperty(_assertThisInitialized(_this), "isSameDay", function (day) {
      var selectedDay = _this.state.selectedDay;
      return _reactDayPicker.DateUtils.isSameDay(selectedDay, day);
    });

    _defineProperty(_assertThisInitialized(_this), "isValidFormat", function (date) {
      var time = _this.props.time;
      var pattern = /^\d{1,4}[.\-/]{1}\d{1,2}[.\-/]{1}\d{1,4}$/;

      if (time) {
        pattern = /^\d{1,4}[.\-/]{1}\d{1,2}[.\-/]{1}\d{1,4}\s{0,1}\d{0,2}([:.])?\d{0,2}$/;
      }

      return pattern.test(date.trim());
    });

    _defineProperty(_assertThisInitialized(_this), "prettifyInputDate", function () {
      var _this$props5 = _this.props,
          value = _this$props5.value,
          dateFormat = _this$props5.dateFormat,
          formatDate = _this$props5.formatDate;

      var momentDate = _moment["default"].utc(value, _moment["default"].ISO_8601);

      var inputDate = formatDate ? formatDate(value) : DateInput.getDate(momentDate, FORMATS.PRETTY_DATE, dateFormat);

      _this.setState({
        inputDate: inputDate
      });
    });

    _defineProperty(_assertThisInitialized(_this), "renderCaptionElement", function (_ref) {
      var date = _ref.date;
      var locale = _this.props.locale;
      return _react["default"].createElement(_yearMonthPicker["default"], {
        date: date,
        onChange: _this.handleYearMonthChange,
        locale: locale
      });
    });

    _defineProperty(_assertThisInitialized(_this), "renderClearValueButton", function () {
      var disabled = _this.props.disabled;
      var className = classPrefix + "-clear-value" + (disabled ? ' disabled' : '');
      return _react["default"].createElement("button", {
        type: "button",
        className: className,
        onClick: _this.handleClearClick,
        disabled: disabled
      }, _react["default"].createElement("span", null, "x"));
    });

    _defineProperty(_assertThisInitialized(_this), "renderDateInput", function () {
      var _this$props6 = _this.props,
          _inputRef = _this$props6.inputRef,
          disabled = _this$props6.disabled,
          formatDate = _this$props6.formatDate,
          inputProps = _this$props6.inputProps,
          showClearValue = _this$props6.showClearValue,
          value = _this$props6.value;
      var inputDate = _this.state.inputDate;
      return _react["default"].createElement(_reactBootstrap.FormGroup, {
        className: classPrefix + "-input-container"
      }, _react["default"].createElement(_reactBootstrap.FormControl, _extends({
        type: "text",
        inputRef: function inputRef(el) {
          _this.input = el;

          _inputRef(el);
        },
        value: inputDate,
        disabled: disabled,
        readOnly: !!formatDate,
        autoComplete: "off"
      }, inputProps, {
        onChange: _this.handleInputChange,
        onFocus: _this.handleInputFocus,
        onBlur: _this.handleInputBlur
      })), showClearValue && value && _this.renderClearValueButton());
    });

    _defineProperty(_assertThisInitialized(_this), "renderCalendar", function () {
      var _this$props7 = _this.props,
          className = _this$props7.className,
          locale = _this$props7.locale,
          time = _this$props7.time,
          value = _this$props7.value,
          inputProps = _this$props7.inputProps,
          inputRef = _this$props7.inputRef,
          disabled = _this$props7.disabled,
          selectedDays = _this$props7.selectedDays,
          showWeekNumbers = _this$props7.showWeekNumbers,
          minutesInterval = _this$props7.minutesInterval,
          showClearValue = _this$props7.showClearValue,
          disabledDays = _this$props7.disabledDays,
          formatDate = _this$props7.formatDate,
          calendarType = _this$props7.calendarType,
          otherProps = _objectWithoutPropertiesLoose(_this$props7, ["className", "locale", "time", "value", "inputProps", "inputRef", "disabled", "selectedDays", "showWeekNumbers", "minutesInterval", "showClearValue", "disabledDays", "formatDate", "calendarType"]);

      var _this$state2 = _this.state,
          dayPickerVisibleMonth = _this$state2.dayPickerVisibleMonth,
          selectedDay = _this$state2.selectedDay;

      var momentDate = _moment["default"].utc(value, _moment["default"].ISO_8601);

      var timeObj = {
        hour: momentDate.hour(),
        minute: momentDate.minute()
      };
      var month = dayPickerVisibleMonth || (typeof selectedDay === 'string' ? undefined : selectedDay);
      return _react["default"].createElement("div", {
        role: "presentation",
        className: classPrefix + "-calendar",
        ref: function ref(el) {
          _this.calendarContainer = el;
        },
        onMouseDown: _this.handleOnOverlayMouseDown
      }, _react["default"].createElement(_reactDayPicker["default"], _extends({}, otherProps, {
        ref: function ref(el) {
          _this.dayPicker = el;
        },
        disabledDays: disabledDays,
        selectedDays: selectedDays || _this.isSameDay,
        localeUtils: _this.localeUtils,
        month: month,
        showWeekNumbers: showWeekNumbers,
        firstDayOfWeek: _this.getFirstDayOfWeek(),
        locale: locale,
        captionElement: _this.renderCaptionElement,
        navbarElement: _navbar["default"],
        onDayClick: _this.handleDayClick
      })), time && _react["default"].createElement(_timePicker["default"], {
        onChange: _this.handleTimePickerChange,
        time: timeObj,
        minutesInterval: minutesInterval
      }));
    });

    var _formatDate = props.formatDate,
        _value = props.value;

    var _momentDate = _moment["default"].utc(_value, _moment["default"].ISO_8601);

    _this.onDocumentClick = _this.onDocumentClick.bind(_assertThisInitialized(_this));

    var _inputDate = _formatDate ? _formatDate(_value) // inputDate: Prettified string shown in input field
    : DateInput.getDate(_momentDate, FORMATS.PRETTY_DATE, props.dateFormat);

    _this.state = {
      /* eslint-disable-next-line react/no-unused-state */
      lastValue: null,
      showOverlay: false,
      // selectedDay: Selected day in calendar (date object)
      selectedDay: DateInput.getDate(_momentDate, FORMATS.DATE_OBJECT, props.dateFormat),
      inputDate: _inputDate
    };
    _this.localeUtils = Object.assign(_moment2["default"], {
      getFirstDayOfWeek: function getFirstDayOfWeek() {
        return _moment["default"].localeData().firstDayOfWeek();
      }
    });
    _this.input = null;
    _this.dayPicker = null; // Used in onBlur handler to determine whether or not blur happened because of a click
    // on the overlay

    _this.mouseClickedOnContainer = false;
    return _this;
  }

  var _proto = DateInput.prototype;

  _proto.componentWillUnmount = function componentWillUnmount() {
    document.removeEventListener('click', this.onDocumentClick);
  }
  /**
   * Fires every time dayPicker is open and document is clicked
   * @param e
   */
  ;

  _proto.render = function render() {
    var _this$props8 = this.props,
        className = _this$props8.className,
        calendarType = _this$props8.calendarType;
    var showOverlay = this.state.showOverlay;

    if (calendarType === 'popup') {
      return _react["default"].createElement(_reactTether["default"], {
        attachment: this.getTetherComponentAttachmentLocation(),
        constraints: [{
          to: 'scrollParent',
          pin: ['top']
        }, {
          to: 'window',
          attachment: 'together'
        }],
        className: classPrefix + " " + className + " " + classPrefix + "-popup-container"
      }, this.renderDateInput(), showOverlay && this.renderCalendar());
    }

    return _react["default"].createElement("div", null, this.renderDateInput(), _react["default"].createElement("div", {
      className: classPrefix + " " + className + " " + classPrefix + "-static-container"
    }, this.renderCalendar()));
  };

  return DateInput;
}(_react["default"].Component);

exports["default"] = DateInput;

_defineProperty(DateInput, "removeInvisibleChars", function (str) {
  return str.replace(/\u200E/g, '');
});

_defineProperty(DateInput, "defaultProps", {
  className: '',
  value: '',
  dateFormat: 'L',
  formatDate: undefined,
  locale: 'en-GB',
  onChange: function onChange() {},
  onDayClick: function onDayClick() {},
  inputProps: {},
  inputRef: function inputRef() {},
  disabled: false,
  selectedDays: null,
  disabledDays: null,
  showOverlay: false,
  showWeekNumbers: true,
  showClearValue: true,
  time: false,
  minutesInterval: 5,
  calendarType: 'popup'
});
