// @opuscapita/react-datetime/lib/cjs/time-picker/time-picker.component.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactBootstrap = require("react-bootstrap");

var _propTypes = _interopRequireDefault(require("prop-types"));

require("./time-picker.scss");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var TimePicker =
/*#__PURE__*/
function (_React$Component) {
  _inheritsLoose(TimePicker, _React$Component);

  function TimePicker(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;

    _defineProperty(_assertThisInitialized(_this), "onChange", function (e) {
      var _Object$assign;

      var _this$props = _this.props,
          time = _this$props.time,
          onChange = _this$props.onChange;

      var oldTime = _extends({}, time);

      var newTime = Object.assign(oldTime, (_Object$assign = {}, _Object$assign[e.target.name] = Number(e.target.value), _Object$assign));
      onChange(newTime);
    });

    _defineProperty(_assertThisInitialized(_this), "getPaddedNumber", function (number) {
      return number < 10 ? "0" + number : number;
    });

    _defineProperty(_assertThisInitialized(_this), "getHourListValues", function () {
      for (var i = 0; i < 24; i += 1) {
        _this.hours.push(i);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "getMinuteListValues", function () {
      var minutesInterval = _this.props.minutesInterval;

      for (var i = 0; i < 60; i += minutesInterval) {
        _this.minutes.push(i);
      }
    });

    _this.hours = [];
    _this.minutes = [];
    return _this;
  }

  var _proto = TimePicker.prototype;

  _proto.componentWillMount = function componentWillMount() {
    this.getHourListValues();
    this.getMinuteListValues();
  };

  _proto.render = function render() {
    var _this2 = this;

    var _this$props2 = this.props,
        _this$props2$time = _this$props2.time,
        minute = _this$props2$time.minute,
        hour = _this$props2$time.hour,
        disabled = _this$props2.disabled;
    return _react["default"].createElement("div", {
      className: "oc-time-picker-container"
    }, _react["default"].createElement(_reactBootstrap.FormControl, {
      name: "hour",
      componentClass: "select",
      value: hour,
      onChange: this.onChange,
      disabled: disabled
    }, this.hours.map(function (h) {
      return _react["default"].createElement("option", {
        key: "hour-" + h,
        value: h
      }, _this2.getPaddedNumber(h));
    })), _react["default"].createElement(_reactBootstrap.FormControl, {
      name: "minute",
      componentClass: "select",
      value: minute,
      onChange: this.onChange,
      disabled: disabled
    }, this.minutes.map(function (m) {
      return _react["default"].createElement("option", {
        key: "minute-" + m,
        value: m
      }, _this2.getPaddedNumber(m));
    })));
  };

  return TimePicker;
}(_react["default"].Component);

exports["default"] = TimePicker;

_defineProperty(TimePicker, "defaultProps", {
  time: {
    hour: 0,
    minute: 0
  },
  minutesInterval: 5,
  disabled: false
});
